incident/mysql/incident_test.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/go-cmp/cmp"
+
+	_ "github.com/go-sql-driver/mysql" // Load MySQL driver
+)
+
+type entry struct {
+	BaseURL, Summary, Category, FullURL, Details string
+}
+
+func checkContents(ctx context.Context, t *testing.T, want []entry) {
+	t.Helper()
+
+	tx, err := testDB.BeginTx(ctx, nil /* opts */)
+	if err != nil {
+		t.Fatalf("failed to create transaction: %v", err)
+	}
+	defer tx.Commit()
+	rows, err := tx.QueryContext(ctx, "SELECT BaseURL, Summary, Category, FullURL, Details FROM Incidents;")
+	if err != nil {
+		t.Fatalf("failed to query rows: %v", err)
+	}
+	defer rows.Close()
+
+	var got []entry
+	for rows.Next() {
+		var e entry
+		if err := rows.Scan(&e.BaseURL, &e.Summary, &e.Category, &e.FullURL, &e.Details); err != nil {
+			t.Fatalf("failed to scan row: %v", err)
+		}
+		got = append(got, e)
+	}
+	if err := rows.Err(); err != nil {
+		t.Errorf("incident table iteration failed: %v", err)
+	}
+	if diff := cmp.Diff(got, want); diff != "" {
+		t.Errorf("incident table: diff (-got +want)\n%s", diff)
+	}
+}
+
+func TestLogf(t *testing.T) {
+	ctx := context.Background()
+	cleanTestDB(ctx)
+
+	checkContents(ctx, t, nil)
+
+	reporter, err := NewMySQLReporter(ctx, testDB, "unittest")
+	if err != nil {
+		t.Fatalf("failed to build MySQLReporter: %v", err)
+	}
+	e := entry{BaseURL: "base", Summary: "summary", Category: "signature", FullURL: "full", Details: "blah"}
+
+	reporter.Log(ctx, e.BaseURL, e.Summary, e.Category, e.FullURL, e.Details)
+	checkContents(ctx, t, []entry{e})
+
+	reporter.Log(ctx, e.BaseURL, e.Summary, e.Category, e.FullURL, e.Details)
+	checkContents(ctx, t, []entry{e, e})
+
+	reporter.Logf(ctx, e.BaseURL, e.Summary, e.Category, e.FullURL, "%s", e.Details)
+	checkContents(ctx, t, []entry{e, e, e})
+}
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	if err := mySQLAvailable(); err != nil {
+		glog.Errorf("MySQL not available, skipping all MySQL storage tests: %v", err)
```

Comment:
It's a shame we can't actually mark the tests skipped, using `t.Skip()`.

Reply:
This is borrowed from the [Trillian MySQL tests](https://github.com/google/trillian/blob/master/storage/mysql/storage_test.go#L284), which allows database creation to be shared across a bunch of tests (and I'd expect there to be more tests here as functionality gets added).

It might be better to create databases (not just rows) per-test, and have boilerplate at the start of each test to do that, but `Test.Main` does seem to be [intended to cover](https://golang.org/pkg/testing/#hdr-Main) this kind of thing: "It is sometimes necessary for a test program to do extra setup or teardown before or after testing."